### PR TITLE
Adding `--with` for CTE

### LINKED ISSUE #684: Unable to run DuckDB queries using FILTER clause on aggregates
I've recently started using jupysql alongside DuckDB to speed up and tidy up my data exploration workflows, which has been great. However, I've recently hit what I think is a bug that's preventing me from using it much more.

If I use a DuckDB query with a [`FILTER` clause](https://duckdb.org/docs/sql/query_syntax/filter) on an aggregate function then jupysql fails to produce a valid final SQL statement and complains about a missing table.

```
CatalogException: Catalog Error: Table with name first_cte does not exist!
Did you mean "pg_type"?
```

Here's an example of a very simple query that reproduced the issue for me on jupysql=0.7.9

```
import duckdb

%load_ext sql
%config SqlMagic.autopandas = True
%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False

%sql duckdb:///:memory:

%%sql --save first_cte --no-execute
SELECT 1 AS column1, 2 AS column2

%%sql --save second_cte --no-execute
SELECT
  sum(column1),
  sum(column2) FILTER (column2 = 2)
FROM first_cte

%%sql
SELECT * FROM second_cte
```

If I run that in a notebook then I get the catalog error above, but if I run the same query in DBeaver then it returns the correct results

CleanShot 2023-06-28 at 10 32 18@2x

Here is the final query that is output by jupysql, which is missing the CTE due to the `FILTER`:

```
final = %sqlrender second_cte
print(final)

WITH
SELECT
  sum(column1),
  sum(column2) FILTER (column2 = 2)
FROM first_cte
```

If I remove the `FILTER` clause then the SQL is generated correctly, though obviously now missing the filter:

```
WITH first_cte AS (
SELECT 1 AS column1, 2 AS column2)
SELECT
  sum(column1),
  sum(column2)
FROM first_cte
```

### PR DESCRIPTION
**Pull request by neelasha23**
## Describe your changes
Adding `--with` again 

## Issue number

Closes #684 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--705.org.readthedocs.build/en/705/

### CODE REVIEW

**Comment by neelasha23** [comment]
@edublancas Please review

File: src/tests/integration/test_generic_db_operations.py [hunk 1]
```diff
@@ -636,10 +637,63 @@ def test_sqlcmd_tables(ip_with_dynamic_db, request):
...
+    "ip_with_dynamic_db",
+    [
+        ("ip_with_postgreSQL"),
+        ("ip_with_mySQL"),
+        ("ip_with_mariaDB"),
+        ("ip_with_SQLite"),
+        ("ip_with_duckDB"),
+        pytest.param(
+            "ip_with_MSSQL",
+            marks=pytest.mark.xfail(reason="issue in capturing the syntax error"),
```
Comment (yafimvo) [inline src/tests/integration/test_generic_db_operations.py:669]: What's the issue? should we open a new issue for it?

**Review by yafimvo — CHANGES_REQUESTED** [review]
Added some comments.

@edublancas @neelasha23 
General thought about the error message:
image

this part: `If using CTEs, you may pass the --with argument explicitly.`
1. What do you think if we add a link to our `--with` documentation
2. Explain in our `--with` docs why and when we should use `--with` (maybe to show this exact case with the error and the solution?)

File: src/tests/integration/test_generic_db_operations.py (hunk 1 above)
Comment (neelasha23) [reply src/tests/integration/test_generic_db_operations.py:669]: I'm having trouble setting up MSSQL locally. and not sure what error is throwing for syntax errors

File: src/tests/integration/test_generic_db_operations.py [hunk 2]
```diff
@@ -636,10 +637,63 @@ def test_sqlcmd_tables(ip_with_dynamic_db, request):
...
+        """%%sql --save first_cte --no-execute
+SELECT 1 AS column1, 2 AS column2
+"""
+    )
+    ip_with_dynamic_db.run_cell(
+        """
+    %%sql --save second_cte --no-execute
+SELECT
+  sum(column1),
+  sum(column2) FILTER (column2 = 2)
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/tests/integration/test_generic_db_operations.py:687]: I think it's supported. I'll be able to debug once mssql works locally.

File: src/tests/integration/test_generic_db_operations.py (hunk 1 above)
Comment (neelasha23) [reply src/tests/integration/test_generic_db_operations.py]: Solved the issue. MSSQL case is also passing now

**Comment by neelasha23** [comment]
> 1. What do you think if we add a link to our `--with` documentation
> 2. Explain in our `--with` docs why and when we should use `--with` (maybe to show this exact case with the error and the solution?)

Added these @yafimvo @edublancas

**Review by edublancas — CHANGES_REQUESTED** [review]
added my feedback in a notebook: [ctes.ipynb.zip](https://github.com/ploomber/jupysql/files/11986956/ctes.ipynb.zip)

**Comment by neelasha23** [comment]
A few doubts from the notebook:

Screenshot 2023-07-08 at 8 00 49 AM
`This is now complaining about first_cte.` : I'm not sure which error? I don't see anything in this cell. Also, this would complain about `first_cte` since sqlglot wouldn't be able to parse the query?



Screenshot 2023-07-08 at 8 01 14 AM
This should work if we pass `--with` while saving the `second_cte` like:

Screenshot 2023-07-08 at 8 07 41 AM

@edublancas

**Comment by edublancas** [comment]
> `This is now complaining about first_cte.` : I'm not sure which error? 

I'm referring to the cell below the "this is now complaining"

> I don't see anything in this cell. Also, this would complain about `first_cte` since sqlglot wouldn't be able to parse the query? 

Ah, you're right. there's no way around it and users will have to pass`--with first_cte --with second_cte`

then the only missing thing is improving the error message.

**Review by yafimvo — COMMENTED** [review]
Looks good! I guess what's left is the error message like @edublancas mentioned.

**Comment by neelasha23** [comment]
> then the only missing thing is improving the error message.

I'm not clear on this part. Are u suggesting to remove the below specific error message completely? This was added as part of [this issue](https://github.com/ploomber/jupysql/issues/229)
Screenshot 2023-07-10 at 5 33 30 PM

**Comment by neelasha23** [comment]
Have modified the error message. However, I still see the old message in [docs](https://jupysql--705.org.readthedocs.build/en/705/compose.html#with-argument).
I have cleared cache.

In local build I see the new one:

Screenshot 2023-07-11 at 7 11 42 PM

Do you also see the old one? @edublancas

**Review by edublancas — APPROVED** [review]

**Pull request merged**